### PR TITLE
GH-1460: Test the dead-letter queue shunt for bad JMS messages.

### PR DESCRIPTION
**Pull request by tbl3rd**
### Purpose

- https://broadinstitute.atlassian.net/browse/GH-1460

### Changes

- Update `clojure` command line options.
- Provide a development testing environment by default.
- Update test inputs to match what is deployed.
- Document more namespaces.
- Reorganize and update `GCS` support to share more code between server and tests.
- Alias test tool namespaces consistently.
- Sort `ns` forms consistently across files.
- Simplify processing of JMS messages.
- Look in the cloud to find possible matches for `::push` file inputs.
- Restrict `:analysisCloudVersion` to files that exist in the cloud.
- Don't copy `::chip` files when staging inputs.  Look for them in `:cloudChipMetaDataDirectory` instead.
- Work around some REPL load order dependencies.
- The jar can now drain the development queue no matter how big `:analysisCloudVersion` is.
- Consolidate timeout, sleep, and authorization functions.
- Remove unused code.
- Break up long lines. Use more consistent names. Tabulate tabular forms.
- Fix JMS message ID handling and the test messages.
- Add an integration test for shunting failing messages to the dead-letter queue.
- Remove the `ptc.unit.jms-test/use-on-prem-idat-paths` test.
- Simplify.

### Review Instructions

- Ignore whitespace changes if you don't want to see them.
- Start in `ptc.util.jms` and `ptc.start` where the important fixes are.
- Then probably look in `ptc.util.environment` because most other changes leverage that.
- Then `ptc.util.gcs` and `ptc.tools.gcs`, and then `ptc.util.misc`, and so on.
- A test is in `ptc.integration.integration-test/forwarding`.

### CODE REVIEW

File: src/ptc/util/jms.clj [hunk 1]
```diff
@@ -123,84 +144,48 @@
...
-  (let [cloud  (env-prefix prefix workflow)
-        result (str/join "/" [cloud "params.txt"])]
-    (misc/gsutil "cp" "-" result :in (jms->params workflow))
+  (let [result (in-cloud-folder prefix workflow "params.txt")]
+    (gcs/gsutil "cp" "-" result :in (jms->params workflow))
     result))
 
-;; See https://broadinstitute.atlassian.net/wiki/spaces/GHConfluence/pages/2853961731/2021-07-28+AoU+Processing+Issue+Discussion
+;; https://broadinstitute.atlassian.net/wiki/spaces/GHConfluence/pages/2853961731/2021-07-28+AoU+Processing+Issue+Discussion
+;; Look first in local filesystem for (input-key workflow).
```
Comment (okotsopoulos) [inline src/ptc/util/jms.clj:156]: Thanks.

File: src/ptc/start.clj [hunk 2]
```diff
@@ -72,61 +73,51 @@
...
+        (if (task! peeked connection)
+          (let [consumed (jms/ednify (consume connection queue))]
+            (log/infof "Task complete, consumed message %s" counter)
+            (if (not (jms/message-ids-equal? peeked consumed))
+              (log/warnf
+               (str/join \space ["Messages differ:"
+                                 (with-out-str
+                                   (pprint (data/diff peeked consumed)))])))
+            (recur (inc counter)))
+          (do (log/errorf "Cannot consume %s" peeked)
```
Comment (okotsopoulos) [inline src/ptc/start.clj:91]: Is there any other info we have at this stage re: the cause of the error?  (The old comment said that this was only reached in testing?)

File: src/ptc/start.clj [hunk 3]
```diff
@@ -72,61 +73,51 @@
...
       (recur counter))))
 
-(defn- message-loop
+(defn ^:private handle-or-dlq
+  "Handle the JMS on CONNECTION or forward to the dead letter queue."
+  [jms connection]
+  (let [dlq    (env/getenv-or-throw "ZAMBONI_ACTIVEMQ_DEAD_LETTER_QUEUE_NAME")
+        bucket (env/getenv-or-throw "PTC_BUCKET_URL")]
+    (try (jms/handle-message bucket jms)
+         true
```
Comment (okotsopoulos) [inline src/ptc/start.clj:101]: What's the point of returning `true` (same within the catch)?

File: src/ptc/util/environment.clj [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  "Define a default dev environment."
+  (:require [clojure.data.json :as json]
+            [clojure.java.io :as io]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [ptc.util.misc :as misc])
+  (:import [java.io FileNotFoundException]))
+
+(def ^:private defaults
+  "Default environment variables to development values for testing."
```
Comment (okotsopoulos) [inline src/ptc/util/environment.clj:11]: Thanks for porting these.

File: test/ptc/integration/integration_test.clj [hunk 5]
```diff
@@ -17,28 +20,55 @@
...
+           (start/produce connection queue "text" properties)
+           (start/listen-and-consume-from-queue task connection queue))
+         (testing "The message was only peeked and can still be consumed"
+           (let [msg (jms/ednify (start/consume connection queue))]
+             (testing "Message is not nil and can be properly read"
+               (is (= @jms-tools/good-jms-message
+                      (select-keys msg [::jms/Properties])))))))))))
+
+(deftest forwarding
+  (testing "forward a bad message to the dead-letter queue"
```
Comment (okotsopoulos) [inline test/ptc/integration/integration_test.clj:52]: Am I right that this integration test has no assertions?  What would the behavior be if the DLQ forwarding didn't work as expected?

File: src/ptc/start.clj (hunk 2 above)
Comment (tbl3rd) [reply src/ptc/start.clj:91]: It's used only for testing.  See the `forwarding` test, and so on.

File: src/ptc/start.clj (hunk 3 above)
Comment (tbl3rd) [reply src/ptc/start.clj:101]: A falsy value terminates the message loop.

File: test/data/test_msg_diff.edn [hunk 6]
```diff
@@ -1,54 +1,57 @@
...
+  :responseRequired false,
+  :timestamp 1597342355447,
+  :producerId "ID:wm8d7-5e4-56463-1597342355067-1:1:1:1",
+  :destination "queue://wfl.broad.pushtocloud.enqueue.dev",
+  :commandId 6,
+  :brokerInTime 1597342355456,
+  :marshalledProperties "org.apache.activemq.util.ByteSequence@3ba3f40d"},
+ :ptc.util.jms/Properties
+ {:action "start",
+  :payload {:Zamboni {:workflow "IlluminaArraysWorkflow",
```
Comment (tbl3rd) [inline test/data/test_msg_diff.edn:26]: Shoulda broken this down too.

File: test/data/test_msg_same.edn [hunk 7]
```diff
@@ -1,54 +1,57 @@
...
+  :responseRequired false,
+  :timestamp 1597342355447,
+  :producerId "ID:wm8d7-5e4-56463-1597342355067-1:1:1:1",
+  :destination "queue://wfl.broad.pushtocloud.enqueue.dev",
+  :commandId 6,
+  :brokerInTime 1597344366123,
+  :marshalledProperties "org.apache.activemq.util.ByteSequence@3ba3f40d"},
+ :ptc.util.jms/Properties
+ {:action "start",
+  :payload {:Zamboni {:workflow "IlluminaArraysWorkflow",
```
Comment (tbl3rd) [inline test/data/test_msg_same.edn:26]: Here again.

File: test/ptc/integration/integration_test.clj (hunk 5 above)
Comment (okotsopoulos) [reply test/ptc/integration/integration_test.clj:52]: Ah, I see now: the assertions are in `ok?`.

Comment (tbl3rd) [reply test/ptc/integration/integration_test.clj:52]: `ok?` asserts.

File: test/ptc/integration/integration_test.clj [hunk 8]
```diff
@@ -17,28 +20,55 @@
...
+                                         (= left right))))]
+        (with-open [connection (start/create-queue-connection
+                                "vm://localhost?broker.persistent=false")]
+          (start/produce connection queue "text" props)
+          (start/listen-and-consume-from-queue handle-or-dlq-just-once
+                                               connection queue)
+          (let [peeked   (jms/ednify (start/peek-message connection dlq))
+                consumed (jms/ednify (start/consume      connection dlq))]
+            (apply ok? (map ::jms/Properties [msg peeked]))
+            (apply ok? (map unbot [peeked consumed]))))))))
```
Comment (tbl3rd) [inline test/ptc/integration/integration_test.clj:74]: Perhaps I should factor the `is` out of `ok?`. Will do.

File: src/ptc/start.clj (hunk 3 above)
Comment (okotsopoulos) [reply src/ptc/start.clj:101]: Ah right, and the point of this improvement is that these bad JMS messages will no longer halt processing.

**Review by okotsopoulos — APPROVED** [review]
Thanks.  I know a bit more about how push to cloud works now.

File: test/ptc/integration/integration_test.clj (hunk 5 above)
Comment (tbl3rd) [reply test/ptc/integration/integration_test.clj:52]: Yeah. Your question prompted me to lift the `is` out to make it more obvious.
I think that is generally good style in tests.

**Pull request merged**